### ISSUE TITLE
The myserious 8-pin chip

### ISSUE THREAD

**Issue by ospilos** [andrei-tatar/imagotag-hack#2]
is in fact just an ordinary SPI flash to store and hold the image data.

**Comment by andrei-tatar** [comment]
Thanks for the info. Do you know the part number/datasheet? Would like to update the docs. Help other hackers :)

**Comment by ospilos** [comment]
Hi,
IIRC it was Winbond 25V10 or more probably W25X10 1Mbit SPI flash in USON package, hence those cryptic markings.
The funny thing was that the CC2510 had not enough RAM to write whole 4K flash pages to the device, so it only contained 8 b&w images. 
EDIT: Looking at the datsheet it seems that it was some other device that was unable to use the whole flash capacity. The 25X01 is programmable by 256 byte chunks, which is definitely possible with CCxx10.  
EDIT1: I digged out my reverse engineering remarks on the board and the flash is indeed W25X10CLUXIG, see https://cz.mouser.com/datasheet/2/949/w25x10cl_revg_021714-1489755.pdf

**Comment by medvm** [comment]
Hello @ospilos, hope you're well!
I've dumped the firmware image from my flash too, but can't decode it to pictures, I used binwalk with various combinations of args, but to no avail. Could you please share your solution?

**Comment by medvm** [comment]
By the way, my version of the board is slightly different from all yours, but the main components are the same. Pictures with captions attached

[Top side](https://user-images.githubusercontent.com/43696206/169647206-84f916aa-ae28-426b-b92b-d52286fcfa55.png)

[Bottom side](https://user-images.githubusercontent.com/43696206/169647176-c0a08957-a5ce-4e15-9a35-317f87bc87d9.png)

**Comment by medvm** [comment]
After numerous attempts, I managed to open the dump as an image, but the image seems to be damaged. It seems that the vertical columns are mixed up in places...
[mixed-up_columns](https://user-images.githubusercontent.com/43696206/169663752-d9917543-5903-41c8-9ebc-0ecd19d09f23.png)

**Comment by medvm** [comment]
Parsed the picture in a normal way. I wonder if the image on the display will be updated after writing a new picture to the flash?..
[res39](https://user-images.githubusercontent.com/43696206/169670860-c0b11286-0eed-432e-b2bd-34a437c9bdfd.png)

**Comment by lols** [comment]
@medvm hi, can you show the way on how you managed to fix the dump?

**Comment by medvm** [comment]
@lols  
If I remember correctly, it was mostly an empirical method. I realised there was definitely something wrong with the on-board winbound flash, so I fixed the dump manually and repeated the remapping sequence for the rest of the dump

**Comment by Kickbut101** [comment]
@medvm 

Hey if you ever get a moment could you please brush up and/or clarify what the challenge here was and how you resolved it? I'm looking at trying to mod and repurpose some of these tags myself and I'm just trying to tie together all the scraps on this repo and maybe even write up a more succinct workflow to help out others. I would really really appreciate it!

**Comment by Loryuwu** [comment]
@Kickbut101 
Hi! How is your version of the project going?
I’m also working on my own implementation, and I’d love to exchange ideas, since I haven’t been able to get RF communication working yet. However, I did manage to store three different images and make them cycle every 6 hours.

I also created my own module for handling the display, because my ESL uses a panel from another provider (Pervasive Displays) instead of GoodDisplays. I’m sharing the link to my project in case you’re interested: [[Here]](https://github.com/Loryuwu/CC2510-ESL?tab=readme-ov-file).

Also, all the documentation and most of the code comments are in Spanish, since it’s my native language and my English is limited (I rely on a translator). Still, I hope my contribution can be helpful.

**Comment by Kickbut101** [comment]
@Loryuwu 

What's funny is I *just* the other day was brushing myself up on where I was at with this project and was looking at all the related repos again.

But no, I have no progression on this. I have lots of imagotags sitting around looking for a good way to use them. I also have a TI Debugging tool on the way but I have almost no idea what I'm doing.

I will browse what you posted, thank you for sharing and don't worry about the translation I can figure it out!

**Comment by Loryuwu** [comment]
@Kickbut101 
Awesome! I’d really appreciate it if you could take a look at my repo and tell me what you think. I feel like I’m pretty much alone working on this, and it’s really hard to find information on how to move forward—especially since the CC2510 is a fairly old chip… but well, it’s what we’ve got.

Right now, my repo has 3 images stored for a 2.66" Pervasive Displays screen. The images are in black and white even though the display also supports red. The code loops through them and switches the image roughly every 6 hours. If you want it to switch every 15 seconds instead, you can replace deep_sleep_seconds(120*6); on line 104 with sleep_ms(500);.
(Yeah, the timing isn’t perfectly accurate yet… I still need to fix that, but it works.)

As I mentioned in another comment, my goal is to get the RF working and then create a module to manage communication with the external memory over SPI and with the NFC chip over I²C. I haven’t started that part yet; right now I’m completely stuck on the RF side.

**Comment by Kickbut101** [comment]
@Loryuwu It looks good, I don't have my TI debugging tool yet, so I cannot attempt to use it. You currently are able to display images that you've setup from the firmware itself right?

~~When you flashed the CC2510 chip were you able to use the exposed pins/pads on the board? or did you have to solder to any points or reroute?~~ derp, should have read closer, you said you had to solder some 
> "Debugger al chip soldando 5 cables, 2 de alimentacion "

**Comment by Loryuwu** [comment]
@Kickbut101 
Hi!
Yes, I’m currently able to display images without any issues. I have a module called EPD, and in EPD.h you can select the display size you want to use. This module is specifically adjusted for Pervasive Displays panels, not GoodDisplays, and I’m not sure whether they are compatible.

To load images, I use [[this]](https://projedefteri.com/en/tools/lcd-assistant/) tool, which converts an image into a hex array with inverted colors, since these displays operate with reversed polarity.

Regarding soldering: the wires I mentioned are only for connecting to the debugger. You can also try using contact methods (pogo pins, adapters, etc.) to avoid soldering directly to the board, since the debugger is only needed to flash the CC2510; after that, it runs fine with just a battery.

In main I have several #if 0 sections that separate the different tests I’m working on: one for the EPD, another for RF, and others for additional features. Feel free to check them out.

Espero y pronto tengas tu Debugger para que le eches un vistazo al script y puedas experimentar también!!

**Comment by Kickbut101** [comment]
You mentioned you do not yet have RF communication working? Is it working at all? Or no?
@Loryuwu

**Comment by Loryuwu** [comment]
@Kickbut101 Yes, that’s correct — I still don’t have RF communication working properly. With the help of AI I’ve made good progress, and my project now has two modules dedicated to handling the radio: rf.h and rf_control.h.

My biggest limitation is that I only have one debugger, so I can’t use SmartRF Studio to configure and test two CC2510 chips directly. Still, I can use it as a scanner: I flash one ESL with a loop that continuously sends packets (powered by a battery), and then with the debugger and another CC2510 I try to listen for those packets in SmartRF.

Behavior of rf.h:
After calling rf_init() and entering the send loop, I use the LEDs for debugging:

Blue LED = radio active

Green LED = send complete

But in this module, rf_send_packet() gets stuck: the blue LED never turns off and the green LED never blinks. In SmartRF I do detect activity, but it’s just unreadable noise.

Behavior of rf_control.h:
This module does complete the transmission. The LEDs behave correctly, and SmartRF receives a packet — it’s corrupted, but at least it’s something structured, unlike the previous module which only produced interference.

In summary:

rf.h → gets stuck; SmartRF receives only noise.

rf_control.h → completes the send; SmartRF receives a corrupted packet.

I'm still investigating why the two modules behave so differently.

**Comment by Loryuwu** [comment]
@Kickbut101 EXELENTES NOTICIAS!!!
perdón que esto lo escriba en español, pero me emociona demasiado, LOGRÉ UNA TRANSMISION SIMPLE POR RF

<img width="491" height="402" alt="Image" src="https://github.com/user-attachments/assets/206f200b-655e-498f-867f-a1076287bf5b" />
en teoría el mensaje es  "Hola mundo desde RF!", por alguna razón las 2 primeras letras se pierden, PERO ES UN ENORME AVANCE, ahora procederé a depurar un poco mejor los módulos y eliminar lo innecesario 

![Image](https://github.com/user-attachments/assets/3e184e51-2e97-47c7-8172-da26c274a084)
en el cuadro verde tenemos un cc2510 sin programa, conectado a través del debugger al SmartRF
en el cuadro azul/purpura tenemos al cc2510 programado para enviar paquetes en bucle, alimentado por sus respectivas baterías y con su cable debug conectado.
